### PR TITLE
Change return type of Tensor::dtype() from ScalarType to TypeMeta

### PR DESCRIPTION
**Pull request by li-roy**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#12766 Change return type of Tensor::dtype() from ScalarType to TypeMeta**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10232118/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #12767 reduce Device to 32bits&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10416051/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13172 use TypeMeta instead of ScalarType in TensorOptions&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D12143282/)

In preparation of using TypeMeta in TensorOptions. 

Differential Revision: [D10232118](https://our.internmc.facebook.com/intern/diff/D10232118/)

### CODE REVIEW

**Comment by goldsborough** [comment]
I agree with @ezyang, operator== is your friend

**Comment by goldsborough** [comment]
@roy-li keep in mind every line you change in the C++ frontend equals one user who gets a confusing build error when they update and costs us github issues and friction with the community

File: aten/src/ATen/core/Tensor.h [hunk 1]
```diff
@@ -160,8 +160,8 @@ class CAFFE2_API Tensor {
   /// Returns a `Tensor`'s layout. Defined in Type.h
   Layout layout() const noexcept;
 
-  /// Returns a `Tensor`'s dtype (`ScalarType`). Defined in Type.h
-  ScalarType dtype() const noexcept;
+  /// Returns a `Tensor`'s dtype (`TypeMeta`). Defined in TensorMethods.h
+  const caffe2::TypeMeta& dtype() const noexcept;
```
Comment (ezyang) [inline aten/src/ATen/core/Tensor.h]: There's no need to return a reference now; it's just a pointer; return it by value.

**Review by ezyang — DISMISSED** [review]
Approved but with comments.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -200,7 +200,7 @@ RegisterOperators reg({
           return [](Stack& stack) {
             at::Tensor a;
             pop(stack, a);
-            push(stack, static_cast<int64_t>(a.dtype()));
+            push(stack, static_cast<int64_t>(a.scalar_type()));
```
Comment (ezyang) [inline torch/csrc/jit/register_prim_ops.cpp]: Maybe TypeMeta can be stored in ivalue? Paging JIT...

File: torch/csrc/jit/script/module.cpp [hunk 3]
```diff
@@ -135,7 +135,7 @@ void Module::to_impl(
     // Use the data's original device or dtype if not supplied here.
     auto new_data = data.to(
         device.value_or(data.device()),
-        dtype.value_or(data.dtype()),
+        dtype.value_or(data.scalar_type()),
```
Comment (ezyang) [inline torch/csrc/jit/script/module.cpp]: This one looks OK.

**Comment by alsrgv** [comment]
@li-roy, @goldsborough,

This change broke the following code: https://github.com/uber/horovod/blob/master/horovod/torch/adapter_v2.cc#L40.  Is there a plan to make it backwards compatible?

**Comment by ezyang** [comment]
@alsrgv The terminal state for DataType is that it will be an open data type, so yeah, that particular use is not going to work, if you're tracking master, use `scalar_type()` method instead. (This will fail if you get any dtype which is not in the ScalarType set.)

**Comment by ezyang** [comment]
(Let us know if we can add anything that will help out your use case.)

**Comment by alsrgv** [comment]
@ezyang, would be great to stabilize the interface or make those breaking changes in bursts.  This is the fourth break after switching to C++ API on Aug 30th.

**Comment by goldsborough** [comment]
@alsrgv Unfortunately I don't think we can promise a lot of stability for the C++ interface in the next few months. We are undertaking large, really important changes to unify ATen with Caffe2 and improve its performance and portability to new platforms (like AMD or custom accelerators) over the long term, and guaranteeing a stable C++ interface would stand in the way of this effort. For this reason we never actually released a stable version of the C++ API, and we currently only release the C++ API in "beta" with the right to make frequent and breaking changes. This is also documented hopefully quite loudly in our docs at https://pytorch.org/cppdocs. In a sense, this should have been clear when you decided to depend on this C++ API (which, just to highlight this, is really the PyTorch backend).

screen shot 2018-11-07 at 4 28 32 pm

I think for your case, it would make the most sense if you pinned particular nightlies of PyTorch and only updated on a monthly-or-so basis. In that case it would be more clear what has changed over the last month, and we could also work together to update your code (which is fortunately not that much). Of course, we do also try our best not to break BC if possible, since many users do depend on the C++ API, and we want to avoid friction unless that friction stands in the way of our progress. Let me know if you like the idea of doing less frequent upgrades of PyTorch.

**Comment by alsrgv** [comment]
@goldsborough, thanks for the comprehensive write-up!  Unfortunately, pinning doesn't work for me because I don't bundle PyTorch, and Horovod is supposed to work with PyTorch that users got from you "today."

We migrated to C++ API primarily C API was being broken as well, and we weren't allowed to make changes, e.g., add Half support.

I suggest we work together on maintaining the compatibility.  I'm not strongly against having a couple of #ifdef's here and there in the code for the sake of supporting PyTorch 0.4.1, 1.0, 1.1, etc., as long as they're used sparingly.

Would you be willing to add `HOROVOD_WITH_PYTORCH=1 pip install git+https://github.com/uber/horovod` to your "C++ API tests" to avoid unintended breaks?

**Comment by goldsborough** [comment]
@alsrgv I think the tricky thing is that we don't actually have a release of the C++ API yet, so there is no version guarding we could do. We're also a bit concerned about the idea that Horovod is supposed to work with PyTorch of "today". If I understand correctly, you are guaranteeing that Horovod works with PyTorch master? But then, isn't any project's master branch supposed to be the land of bold, unconstrained breaking changes? I don't think we can ever guarantee any stability for the C++ API on master, because that's fundamentally really not what master is intended for.

We discussed adding the horovod build to our CI, but we don't like the idea of particularly favoring one partner repository to run in our CI. If other folks see that we added some kind of higher-class support for one project, I don't see it taking long until lots of other projects expect us to run their build in our CI and ensure backwards compatibility and that will just slow us down a whole deal.

I still think you should just pin PyTorch to some particular nightly and have users depend on that version. I feel like depending on released (or in this case at least fixed) versions of projects is a well established practice of ensuring stability, and depending on master is always dangerous and comes with responsibility for the dependee to keep up to date with the changes of upstream.

**Comment by alsrgv** [comment]
@goldsborough, the reality is that you keep adding new great features to PyTorch, and subsequently users keep upgrading to latest version of PyTorch which may be incompatible with plugins like Horovod.  Horovod only supporting certain pin of PyTorch master won't work, because users won't be able to use those new great features (and they're typically very impatient!).

If making a formal release of C++ API is not an option at this point, the ask is to make breaking changes in bursts, rather than trickling through.  This way, plugin authors would have to make less frequently "catch up" releases.

**Comment by soumith** [comment]
@alsrgv between 0.4.1 and 1.0 stable there is a big gap, close to 5 months. We also urged users to try 1.0 preview. This was a one-time situation as we transitioned to our first stable release. Usually our release cycle is 3 months, and we always urge users to be on stable.
For everyone in this conversation, after December 3rd we will be in a happy stable shape, only having downstream packages use stable releases again.

**Comment by alsrgv** [comment]
@soumith, great!  It'd still be great if breaking changes could be batched - even before 0.4.1 I knew a lot of researchers who were sitting on the master branch. :-)

**Pull request closed without merging**